runtime/common/src/account_conversion.rs

```diff
@@ -11,93 +11,85 @@
+	//    here.
+	let tag = b"EVM";
+	let mut bytes = [0; 32];
+	bytes[0..20].copy_from_slice(&address);
+	bytes[20..28].copy_from_slice(&chain_id.to_be_bytes());
+	bytes[28..31].copy_from_slice(tag);
+	AccountId::new(bytes)
 }
 
+/// Common converter code for translating from eth accounts
+pub struct AccountMapping<R>(sp_std::marker::PhantomData<R>);
+
 // Implement EVM account conversion using our shared conversion code
-impl<R, XcmConverter> AddressMapping<AccountId> for AccountConverter<R, XcmConverter>
-where
-	R: pallet_evm_chain_id::Config,
-{
+impl<R: pallet_evm_chain_id::Config> AddressMapping<AccountId> for AccountMapping<R> {
 	fn into_account_id(address: H160) -> AccountId {
 		let chain_id = pallet_evm_chain_id::Pallet::<R>::get();
-		Self::convert_evm_address(chain_id, address.0)
+		convert_evm_address(chain_id, address.0)
 	}
 }
 
-// Implement liquidityPools account conversion using our shared conversion code
-impl<R, XcmConverter> Convert<DomainAddress, AccountId> for AccountConverter<R, XcmConverter> {
+/// Common converter code for translating accounts across different
+/// domains and chains.
+pub struct AccountConverter;
+
+impl Convert<DomainAddress, AccountId> for AccountConverter {
 	fn convert(domain_address: DomainAddress) -> AccountId {
 		match domain_address {
 			DomainAddress::Centrifuge(addr) => AccountId::new(addr),
-			DomainAddress::EVM(chain_id, addr) => Self::convert_evm_address(chain_id, addr),
+			DomainAddress::EVM(chain_id, addr) => convert_evm_address(chain_id, addr),
 		}
 	}
 }
 
-impl<R, XcmConverter> Convert<(Domain, [u8; 32]), AccountId> for AccountConverter<R, XcmConverter> {
+impl Convert<(Domain, [u8; 32]), AccountId> for AccountConverter {
 	fn convert((domain, account): (Domain, [u8; 32])) -> AccountId {
 		match domain {
 			Domain::Centrifuge => AccountId::new(account),
 			// EVM AccountId20 addresses are right-padded to 32 bytes
 			Domain::EVM(chain_id) => {
 				let mut bytes20 = [0; 20];
 				bytes20.copy_from_slice(&account[..20]);
-				Self::convert_evm_address(chain_id, bytes20)
+				convert_evm_address(chain_id, bytes20)
 			}
 		}
 	}
 }
 
-impl<R, XcmConverter> TryConvert<v3::MultiLocation, AccountId> for AccountConverter<R, XcmConverter>
-where
-	XcmConverter: xcm_executor::traits::Convert<v3::MultiLocation, AccountId>,
-{
-	type Error = v3::MultiLocation;
-
-	fn try_convert(location: v3::MultiLocation) -> Result<AccountId, Self::Error> {
-		// Try xcm logic first
-		match XcmConverter::convert_ref(location).ok() {
-			Some(acc) => Ok(acc),
-			None => {
-				// match EVM logic
-				match location {
-					v3::MultiLocation {
-						parents: 0,
-						interior:
-							v3::Junctions::X1(v3::Junction::AccountKey20 {
-								network: Some(v3::NetworkId::Ethereum { chain_id }),
-								key,
-							}),
-					} => Ok(Self::convert_evm_address(chain_id, key)),
-					_ => Err(location),
-				}
+pub fn location_to_account<XcmConverter: ConvertLocation<AccountId>>(
```

Comment:
Nothing forces you now to use `TryConvert`, so now the implementation is a standalone function. Now it returns an `Option` instead of a `Result`. But note that the previous error used in the result was the same input given. So, actually, there was no extra information on it.

Reply:
This method is the same as the previous `TryConvert` implementation for `AccountConverter`. 

The method can not be implemented in the same exact way because the requirement `xcm_executor::traits::Convert` no longer exists. Also, no associated type demands this behavior under a `TryConvert`, so we do not need this kind of implementation now.

> Is this tested with out XCM logic?

As it worked before. And before: it was not tested. But we can do it as new work.

> AFAIR the XCM account conversion logic is a waterfall of converters where the location is always returned from the failing previous try.

This remains exactly the same as before; some traits have changed, but the "waterfall effect" is there.

> Also I do not like to have separate functions for all of this. The main reason to have one structure is to not ensure all are using the same logic. This gets lets obvious with having methods here.

I have a strong opinion of having separated:

- From the perspective of someone who doesn't know how this works, forcing them to pass a `Runtime` and a `LocationToAccountId` every time for a method that does not require them confuse a lot. And make some simpler things that should be easy to generalize very difficult. And it becomes a nightmare if, in future updates, we need to modify those traits again.

- All methods/types are still using the same structure, all of them end up to `convert_evm_address()`. It's ensured by compile time that all of them use the same logic. Nothing has changed related to it. So there is no any stability problem regarding this.

- i.e. If I need to convert a evm address previously I need: 
```rust
AccountConverter::<centrifuge_runtime::Runtime, centrifuge_runtime::xcm::LocationToAccountId>::convert_evm_address(chain_id, addr)
```

when I actually need to just call:
```rust
convert_evm_address(chain_id, addr)
```

I do not need to look for a `LocationToAccountId` or belong to a `Runtime` to call it, where sometimes it is very difficult or impossible to get it without passing this type in other places.

Reply:
I disagree that splitting the `AccountConverted` into `AccountMapping` and another `AccountConverter` is a good idea. IMO, this gives the wrong perception that one can do mappings differently. The main reason this should be under one umbrella is so ensure it is easy to spot whenever the wrong conversion is used. 

Same for the `const Account = convert_evm_address()`. Again, hard to track where which conversion is used. Also implies I can use another conversion in tests or for another `AccountId`.

IMO we should try to get rid of the second generic for the `LocationToAccountId`. Passing the Runtime is basically susbtrate standard and should not be anymore confusing than anywhere else.

Reply:
> The main reason this should be under one umbrella is so ensure it is easy to spot whenever the wrong conversion is used.

> hard to track where which conversion is used

Could you extend this? Or put an example? I'm not getting the extra benefits it offers to have these under the same type. What do you want is a kind of namespace to look for it faster?

Regarding account mapping. It only exists to satisfy the following trait:

```rust
impl pallet_evm::Config for Runtime {
    type AddressMapping = AccountMapping<Runtime>;
    //...
}
```

It's not like the user can choose between `AccountMapping` or `AccountConverter`, both are types that are implemented just to satisfy the trait demands. And all methods have different signatures.

Reply:
About adding a generic that is not used in the implementation. What it scares me is the following:
1. I call `AccountConverter::<Altair>::convert_evm_address()`. Nevertheless, it should be Centrifuge, I made a mistake, but the compiler does not tell me anything because the type is not used, so it works with any type.
2. After an update, `convert_evm_address()` make use of the runtime to get some property, e.i, a `chain_id` or a `parachain_id`. Everything compiles, but the line above will return a different value than expected without noticing it.

Actually, I've already found several of these issues during this migration: i.e.: https://github.com/centrifuge/centrifuge-chain/pull/1756/files#diff-f2256017c4e606a30f53708578da46675512cc1300be6542ed5f77e58af0d6dcR804

Reply:
I do not want to be obstinate or close mind 😆, so I'll anyways do a PR with the changes I think you expect, and we can see how it changes in the database.

Reply:
Yes, true the namespacing in this lib here is clear. But once you import you only see the function. I will not block this change, but I think having one struct that implements all conversions and is used everwhere accross the codebase has a huge benefit. Everywhere where there is an account conversion needed, you use this struct. 

Ensuring that not using it definitely is spotted easily, ensuring all conversions are stable. 

Regarding using the wrong runtime, I agree that is a thing. But the struct is mostly used in the runtime configuration where it is impossible to be missused in that sense. Still for the node that is an issue, I agree.

Reply:
I think here we can have what both want: https://github.com/centrifuge/centrifuge-chain/pull/1806